src/CommunityToolkit.Maui/ImageSources/GravatarImageSource.shared.cs

```diff
@@ -0,0 +1,239 @@
+	internal int ParentHeight => (int)GetValue(ParentHeightProperty);
+
+	/// <summary>Gets or sets the parent width.</summary>
+	internal int ParentWidth => (int)GetValue(ParentWidthProperty);
+
+	/// <summary>Gets or sets the image size.</summary>
+	int GravatarSize
+	{
+		get => gravatarSize;
+		set => gravatarSize = Math.Clamp(value, 1, 2048);
+	}
+
+	/// <summary>Returns the Uri as a string.</summary>
+	/// <returns>String of the URI.</returns>
+	public override string ToString()
+	{
+		return $"Uri: {Uri}, Email:{Email}, Size:{GravatarSize}, Image: {DefaultGravatarName(Image)}, CacheValidity:{CacheValidity}, CachingEnabled:{CachingEnabled}";
+	}
+
+	/// <summary>On parent set.</summary>
+	protected override void OnParentSet()
+	{
+		object element = Parent;
+		base.OnParentSet();
+		if (element is not VisualElement parentElement || parentElement is null)
+		{
+			GravatarSize = defaultSize;
+			return;
+		}
+
+		SetBinding(ParentWidthProperty, new Binding(nameof(VisualElement.Width), BindingMode.OneWay, source: parentElement));
+		SetBinding(ParentHeightProperty, new Binding(nameof(VisualElement.Height), BindingMode.OneWay, source: parentElement));
+	}
+
+	static string DefaultGravatarName(DefaultImage defaultGravatar)
+			=> defaultGravatar switch
+			{
+				DefaultImage.FileNotFound => "404",
+				DefaultImage.MysteryPerson => "mp",
+				_ => $"{defaultGravatar}".ToLower(),
+			};
+
+	static string GetMd5Hash(ReadOnlySpan<char> str)
+	{
+		using var md5 = MD5.Create();
+		Span<byte> hash = md5.ComputeHash(Encoding.UTF8.GetBytes(str.ToArray()));
+		return BitConverter.ToString(hash.ToArray(), 0, hash.Length).Replace("-", string.Empty, StringComparison.OrdinalIgnoreCase).ToLowerInvariant();
+	}
+
+	static void OnDefaultImagePropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleNewUriRequested(gravatarImageSource.Email, (DefaultImage)newValue);
+	}
+
+	static void OnEmailPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleNewUriRequested((string?)newValue, gravatarImageSource.Image);
+	}
+
+	static void OnSizePropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		if (newValue is not int intNewValue || intNewValue <= -1)
+		{
+			return;
+		}
+
+		if (gravatarImageSource.GravatarSize == 0)
+		{
+			gravatarImageSource.GravatarSize = intNewValue;
+			return;
+		}
+
+		gravatarImageSource.GravatarSize = Math.Min(intNewValue, gravatarImageSource.GravatarSize);
+		gravatarImageSource.HandleNewUriRequested(gravatarImageSource.Email, gravatarImageSource.Image);
+	}
+
+	async Task<Stream> DownloadStreamAsync(Uri? uri, CancellationToken cancellationToken)
```

Comment:
maybe we can move it to Extensions as well. it is rather abstract and may be useful not only in GravatarImageSource

Reply:
Not a bad idea. Are you thinking an extension method on `HttpClient`?

Reply:
HttpClient or Uri

Reply:
@VladislavAntonyuk  I've added a `HttpClientExtensions.shared.cs` and changed `GravatarImageSource` to use this.  Let me know if it's good with you. (We can obviously add more extension methods later, as needed).